src/crimson/net/ProtocolV2.cc

```diff
@@ -0,0 +1,1603 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "ProtocolV2.h"
+
+#include "include/msgr.h"
+#include "include/random.h"
+
+#include "crimson/auth/AuthClient.h"
+#include "crimson/auth/AuthServer.h"
+
+#include "Config.h"
+#include "Dispatcher.h"
+#include "Errors.h"
+#include "Socket.h"
+#include "SocketConnection.h"
+#include "SocketMessenger.h"
+
+using namespace ceph::msgr::v2;
+
+namespace {
+
+seastar::logger& logger() {
+  return ceph::get_logger(ceph_subsys_ms);
+}
+
+seastar::future<> abort_in_fault() {
+  throw std::system_error(make_error_code(ceph::net::error::negotiation_failure));
+}
+
+seastar::future<> abort_in_close() {
+  throw std::system_error(make_error_code(ceph::net::error::connection_aborted));
+}
+
+inline void expect_tag(const Tag& expected,
+                       const Tag& actual,
+                       ceph::net::SocketConnection& conn,
+                       const char *where) {
+  if (actual != expected) {
+    logger().error("{} {} received wrong tag: {}, expected {}",
+                   conn, where,
+                   static_cast<uint32_t>(actual),
+                   static_cast<uint32_t>(expected));
+    abort_in_fault();
+  }
+}
+
+inline seastar::future<> unexpected_tag(const Tag& unexpected,
+                                        ceph::net::SocketConnection& conn,
+                                        const char *where) {
+  logger().error("{} {} received unexpected tag: {}",
+                 conn, where, static_cast<uint32_t>(unexpected));
+  return abort_in_fault();
+}
+
+} // namespace anonymous
+
+namespace ceph::net {
+
+ProtocolV2::ProtocolV2(Dispatcher& dispatcher,
+                       SocketConnection& conn,
+                       SocketMessenger& messenger)
+  : Protocol(2, dispatcher, conn, messenger) {}
+
+ProtocolV2::~ProtocolV2() {}
+
+void ProtocolV2::start_connect(const entity_addr_t& _peer_addr,
+                               const entity_type_t& _peer_type)
+{
+  ceph_assert(state == state_t::NONE);
+  ceph_assert(!socket);
+  conn.peer_addr = _peer_addr;
+  conn.target_addr = _peer_addr;
+  conn.peer_type = _peer_type;
+  // TODO: lossless policy
```

Comment:
shall we query the messenger for the policy that applies to given `peer_type` ?

Reply:
done in https://github.com/ceph/ceph/pull/26710/commits/d481448fe1439f5eb485d77265b44f974d6ef707